### ISSUE TITLE
TF 2.4.0 with disable_eager_execution twice slowlier than 2.3.1 with disable_eager_execution

### ISSUE THREAD

**Issue by Marmotte06** [tensorflow/tensorflow#45888]
I just upgraded from TF 2.3.1 to TF 2.4.0 and my tf.keras model is at least twice slowlier than it was. Is anyone encountering the same problem ?

I am on Windows 10 x64 2004
Python 3.8.6
GPU RTX 2070 8 Gb RAM
CPU i9 9900 K
64 Gb RAM, 1Tb SSD
CUDA 11.0
CuDNN 8.0.2
Numpy 1.19.3
and I disabled Keras eager execution (disabled since I use TF 2.0) with the statement: tensorflow.compat.v1.disable_eager_execution()

The execution console gives:
--------------------------------------------------------
2020-12-20 17:42:12.138275: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2020-12-20 17:49:15.949788: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2020-12-20 17:49:15.951856: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library nvcuda.dll
2020-12-20 17:49:15.991681: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2070 computeCapability: 7.5
coreClock: 1.62GHz coreCount: 36 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.29GiB/s
2020-12-20 17:49:15.991885: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2020-12-20 17:49:16.099671: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublas64_11.dll
2020-12-20 17:49:16.099749: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublasLt64_11.dll
2020-12-20 17:49:16.146047: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2020-12-20 17:49:16.181370: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2020-12-20 17:49:16.369302: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2020-12-20 17:49:16.412993: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusparse64_11.dll
2020-12-20 17:49:16.426034: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudnn64_8.dll
2020-12-20 17:49:16.426310: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
Model: "model"
Layer (type)                    Output Shape         Param #     Connected to        
__________________________________________________________________________________________________
input_1 (InputLayer)            [(None, 8)]          0         
__________________________________________________________________________________________________                      
input_2 (InputLayer)            [(None, 22)]         0                               
__________________________________________________________________________________________________
embtri10 (Embedding)            (None, 8, 16)        235584      input_1[0][0]       
__________________________________________________________________________________________________
fcpos (Dense)                   (None, 19)           437         input_2[0][0]       
__________________________________________________________________________________________________
fctri (FC_SYMETRIQUE)           (None, 46)           5888        embtri10[0][0]      
__________________________________________________________________________________________________
concatenate (Concatenate)       (None, 65)           0           fcpos[0][0]              fctri[0][0]         
__________________________________________________________________________________________________
line (Dense)                    (None, 1)            65          concatenate[0][0]   
__________________________________________________________________________________________________
Total params: 241,974
Trainable params: 241,974
Non-trainable params: 0

2020-12-20 17:49:16.988484: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-12-20 17:49:16.989498: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2070 computeCapability: 7.5
coreClock: 1.62GHz coreCount: 36 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.29GiB/s
2020-12-20 17:49:16.989582: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2020-12-20 17:49:16.991351: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublas64_11.dll
2020-12-20 17:49:16.991616: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublasLt64_11.dll
2020-12-20 17:49:16.991879: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2020-12-20 17:49:16.992133: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2020-12-20 17:49:16.992402: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2020-12-20 17:49:16.992655: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusparse64_11.dll
2020-12-20 17:49:16.992919: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudnn64_8.dll
2020-12-20 17:49:16.993210: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
2020-12-20 17:49:17.759695: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-12-20 17:49:17.759774: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0
2020-12-20 17:49:17.760494: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N
2020-12-20 17:49:17.763105: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6637 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2070, pci bus id: 0000:01:00.0, compute capability: 7.5)
2020-12-20 17:49:17.763980: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2020-12-20 17:49:17.776001: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)
2020-12-20 17:49:17.940776: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublas64_11.dll
2020-12-20 17:49:18.573803: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublasLt64_11.dll
__________________________________________________________________________________________________
Epoch 1/24
10165/10165 - 95s - loss: 0.7623 - mae: 0.0554 - mse: 0.0053
Epoch 2/24
10165/10165 - 73s - loss: 0.7172 - mae: 0.0518 - mse: 0.0045
etc.
__________________________________________________________________________________________________
whereas with TF 2.31 I use to have, for exactly the same model and data:
Epoch 1/24
10165/10165 - 34s - loss: 0.7643 - mae: 0.0555 - mse: 0.0053
Epoch 2/24
10165/10165 - 34s - loss: 0.7183 - mae: 0.0519 - mse: 0.0046
etc.

My model is compiled with:
OPTIM = Adam(lr=LR_ADAM)
MODELE.compile(optimizer=OPTIM, loss=MSE_BCE_LOSS, metrics=['mae', 'mse'])

And the Fit statement:
MODELE.fit(x=BATCH(T_JEUX, T_PARAMS, T_SCORES, BATCH_SIZE),
               epochs=NB_EPOCHS_ADAM, shuffle=False, verbose=2)

With BATCH being a Sequence

Olivier

**Comment by Marmotte06** [comment]
And I add that my GPU utilization has dropped to only 5% !

**Comment by rmothukuru** [comment]
@Marmotte06,
In order to expedite the trouble-shooting process, please provide a code snippet or a Colab Gist to reproduce the issue reported here. Thanks!

**Comment by Marmotte06** [comment]
@rmothukuru 
I gave you access to the Python/Keras code in my GitHub repository, but the data files you would need to run it are far too big to be uploaded on GitHub. Don't know how to share them.

**Comment by rmothukuru** [comment]
@Marmotte06,
Will it be possible for you to replicate the issue with a simple stand alone code. Thanks!

**Comment by cr0m3r0** [comment]
You can speed up the process enabling XLA and jit, and disabling CUDA LAUNCH BLOCKING, but may present some memory allocations problems and random freezes. Use the following before creating variables and models.

```
import os
import tensorflow as tf 

os.environ['TF_DEVICE_MIN_SYS_MEMORY_IN_MB'] = '1050'
os.environ['TF_XLA_FLAGS']='--tf_xla_enable_xla_devices --tf_xla_auto_jit=fusible --tf_xla_cpu_global_jit --tf_xla_always_defer_compilation=false --tf_xla_enable_lazy_compilation=true'

os.environ['TF_ENABLE_GPU_GARBAGE_COLLECTION']='false'
os.environ['TF_FORCE_GPU_ALLOW_GROWTH']='true'

os.environ['CUDA_LAUNCH_BLOCKING'] = '0'

print('jit:',tf.config.optimizer.get_jit())
tf.config.optimizer.set_jit(True)
print('jit:',tf.config.optimizer.get_jit())
```

**Comment by Marmotte06** [comment]
@rmothukuru,

What I could do is a special version of my program that bypasses all the data prep phase and load a pre-prepared data set so it immediately runs the Model training. And I could make the data set smaller by a factor 10 or twenty to make it smaller while still having measurable epoch times. 
Would that help you understand what's going on ?

Apart from that, I looked at the change log between TF 2.3.1 and TF 2.4 and didn't find anything that would break my code. The only specificities of my code is that I have defined a custom Keras layer and a custom loss function, which bot use TF functions such as **tensor slicing (inputs[:, 0]), tf.concat, tf.matmul, tf.square, tf.cast, tf.sign, tf.nn.sigmoid_cross_entropy_with_logits, tf.reduce_mean**.
**Do you think that could be a problem when compiling the model with TF 2.4 ?**
I used to use the backend K. function is a previous version, but they seem to have disappeared now that Keras only supports the TF backend.

**Comment by Marmotte06** [comment]
@cr0m3r0

Thanks for the suggestion, will try. But does this explain why the same model takes 2+ more time to train with TF 2.4 than it used to with TF 2.3 ?
When reading the TF 2.4 change log I noticed some issues with the XLA jit compiler on Ampere cards (I am still on a Turing card), but this seems to affect the startup time rather than the epoch time, isn't it ?

**Comment by Marmotte06** [comment]
I just made another test with a simpler Keras model, just one embedding layer and two dense layers, no custom layer or custom loss.

The result is even worse, each training epoch takes 3 times (x 3) longer with TF 2.4.0 than it was taking with TF 2.31. Same model, same data, same machine.

So it's not a problem with the model, it's a problem somewhere between the SW stack and the RTX 2070.
Tomorrow I'll check I have the most recent NVidia device driver.

**Comment by Marmotte06** [comment]
Ok, I might have a clue on what's going on:

When upgrading from 1.5 to 2.0 I experienced a severe training performance impact due to Eager execution, which I resolved with tensorflow.compat.v1.disable_eager_execution()
That workaround worked fine until now (TF 2.3.1)

This morning I commented out the disable_eager_execution() instruction and... got the exact same bad performance with TF 2.4 than when that instruction is present.

So my guess is that I am suffering again the penalty of Eager execution, even though I am trying to disable it (I do not need Eager execution).

**Hence that performance issue might actually be a bug, i.e. tensorflow.compat.v1.disable_eager_execution() doesn't work anymore.**

Or, is there a new API to disable Eager execution and avoid the penalty of it ?

**Comment by tomerk** [comment]
Hi @Marmotte06,

can you share a colab gist of the simple (3 layer) model you mentioned above? We know about some gpu embedding-placement-related performance issues on TF 2.x vs TF 1.x, but these shouldn't have arisen between TF 2.3.1 & TF 2.4. (The embeddings get placed on CPU even in cases where they should be placed on gpu for performance reasons).

Or are you saying in TF 2.3.1 you used to disable eager execution to get your performance and now you aren't disabling it?

Either way, can you give the OneDeviceStrategy a shot? (Make your keras model inside a https://www.tensorflow.org/api_docs/python/tf/distribute/OneDeviceStrategy that specifies the gpu). That should force the embeddings to be placed on gpu (tf.device won't unfortunately), and you can see if your performance improves then.

**Comment by Marmotte06** [comment]
@tomerk 

On TF 2.3.1 I was disabling eager execution to keep the performance I had with TF 1.5 (otherwise I got a x2 training time increase)
What I noticed with TF 24 is that I do have the x2 training time increase anyway, eager execution on or off. Which let me think the eager execution trick doesn't work anymore.

I just tried the OneDeviceStrategy, including the -functional- model layer definitions and model compilation in the strategy scope, but not the fit instruction (I also tried to put the fit() in).

I used the following statement:
strategy = tf.distribute.OneDeviceStrategy(device="/gpu:0")

since when starting TF tells on the console:
2021-01-08 10:48:49.623571: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6637 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2070, pci bus id: 0000:01:00.0, compute capability: 7.5)

Unfortunately I get the following assertion error:

  File "C:\Users\Olivier\AppData\Roaming\Python\Python38\site-packages\tensorflow\python\keras\engine\training_v1.py", line 789, in fit
    return func.fit(
  File "C:\Users\Olivier\AppData\Roaming\Python\Python38\site-packages\tensorflow\python\keras\engine\training_distributed_v1.py", line 615, in fit
    dataset = model._distribution_standardize_user_data(
  File "C:\Users\Olivier\AppData\Roaming\Python\Python38\site-packages\tensorflow\python\keras\engine\training_v1.py", line 2195, in _distribution_standardize_user_data
    assert isinstance(x, dataset_ops.DatasetV2)
AssertionError

Any clue ?

**Comment by Marmotte06** [comment]
@tomerk

I made some progress. The "training_v" mentions in the error above let me think my compat.v1 import disable_eager_execution could be the cause of the error. I disabled that statement and the training started, even though I now get the following warning:

2021-01-08 13:18:37.888518: W tensorflow/core/grappler/optimizers/data/auto_shard.cc:656] In AUTO-mode, and switching to DATA-based sharding, instead of FILE-based sharding as we cannot find appropriate reader dataset op(s) to shard. Error: Did not find a shardable source, walked to a node which is not a dataset: name: "FlatMapDataset/_2"
Don't know what that means and how to fix it, but I can see some performance improvement:

With 2.3.1 and Eager execution disabled I got 37 seconds / epoch
With 2.4 off-the-shelf I got 77 seconds / epoch
And with 2.4 and OneDeviceStrategy I get 62 sec / seconds.

So, I am not back to the 2.3.1 performance, but it's improving.

And...  it seems I get a slightly better training accuracy:
loss: 0.6421 - mae: 0.0452 - mse: 0.0034   instead of:    loss: 0.6434 - mae: 0.0453 - mse: 0.0035  after the same number of epochs.
My model is quite reproducible, so that's not an artifact. Any clue for this improvement ? I imagine doing the embedding training on the GPU might give some precision improvement over a mixed CPU/GPU training ?

**Comment by tomerk** [comment]
Yeah the distribution strategy probably won't work when you disable eager execution (effectively entering TF1 graphs/execution).

The remaining performance gap does sound like something we need to look into. As requested above can you share a colab gist of your simpler three-layer model repro?

It would also be really helpful if you could share TF profiler traces from these different runs in tf 2.3.1 & tf 2.4:
https://www.tensorflow.org/guide/profiler

**Comment by Marmotte06** [comment]
Ok, I'll have a look at the profiler. However I don't have the 2.3.1 anymore. But maybe I'll roll back to it to get back my original performances until the 2.4 improves.

I really like your idea of putting the embeddings on the GPU (and the improved accuracy), but I get tons of those "sharding" warnings that are very disturbing. I did some research and found I would have to change all my numpy arrays to Datasets to get rid of those warning, but that would probably change all my Sequence logic, so I don't want to go that path.
I liked Keras for it's simplicity, but from version to version it is getting to a level of complexity that makes me wonder whether that's still the right choice. Maybe I should have a look to PyTorch.

**Comment by Marmotte06** [comment]
I also have this warning from TF 2.4
2021-01-08 13:18:37.898614: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)

Does this have any performance impact ?

**Comment by Marmotte06** [comment]
In an effort to turnaround the problem, I rewrote my Sequence generator logic into a plain tf.data.Dataset logic.
After struggling a bit with the the necessary steps, parameters, options, etc. it eventually works fine.
The good news is that don't have lengthy warning messages anymore and that the new code is much cleaner with the Dataset directly created from my Numpy arrays, all the code for the Sequence generator disappears.
The bad news is that I still have 64 seconds / epoch, i.e. twice the time I got with 2.3.1.

The next step is therefore to take the time to use the Profiler as suggested by @tomerk 

Apart from the inconvenience of a doubled compute time, I definitively confirm that I get a better accuracy by moving the embeddings to the GPU, hence I am happy anyway with the move and will convert all my Sequences to Datasets.

**Comment by Marmotte06** [comment]
After struggling to install TensorBoard and the Profiler on Windows 10 (the cubti DLL has changed name with CUDA 11), I could eventually run a profiling session.
There is a lot of Profiling information, I don't really know where to look.
But the following figure strikes me, even though I suspected it given my 4% GPU utilization:
TF Op Placement:
    Host: 95.7%
    Device: 4.3%
In the bottleneck analysis I could find this:
Summary of All Input Pipelines
Host              Input Pipeline       Min (us)       Avg (us)       Max (us)      # call    # slow calls
PC-OLIVIER      Device:0                2                  2                 12              21             0
PC-OLIVIER      Host:0                6,193          7,842           10,094           21           21
followed by this recommendation:
 1. Check the locality of a host and input data. Ideally, they should be in the same cell (or very close, like the same region).
2. Parallelize reading from this dataset source

What does that mean, knowing that all my data is loaded in RAM ?    (I have 64 GB of RAM and an i9 9900K)

In the statistics I can see:
                                                         #Occurrences     Total self-time (us)    Total self-time on Host (%)
Dataset                                                  156,589             163,071                          97.7%
FlushSummaryWriter                                   1                    3,624                            2.2%
MultiDeviceIteratorFromStringHandle        21                     154                              0.1%
MultiDeviceIteratorGetNextFromShard       21                      78                               0%
LogicalAnd                                                  1                       20                               0%
WriteSummary                                            1                        7                                0%
Identity                                                       1                        0                                0%

Any clue ?

**Comment by Marmotte06** [comment]
I did additional testing:
The above figure (64 sec / epoch) was obtained with a dataset created with tf.data.Dataset.from_tensor_slices(), knowing that all my training data is available in RAM in Numpy arrays.
However, my Numpy arrays a quite big, up to tens of Gigabytes (the biggest datasets saturate my 64 Gb RAM), hence I tried the recommended approach with a Python Generator that yields rows of my Numpy arrays, one at a time, and tf.data.Dataset.from_generator() of that Generator, plus repeat(), batch(4096) and prefetch(autotune).

The result is just terrible, each epoch now takes forever and my GPU usage is ZERO % !

What's going on ???

**Comment by taesookim0412** [comment]
After upgrading to 2.4 while still using a Turing card, I initially found that my average inferencing algorithms increased in time required by 150%, increasing over time, only after running for at least 12 hours. This average was over the entire period, so the time required may have increased to 200%. I am using a turing card with cuda 11. 

I noticed that my GPU usage kept dropping really badly. Instead of when I restart and having full usage with the occasional spike it was probably only utilizing 80% on average and was dropping at very low utilization. This leads me to believe that XLA wasn't going very well.

> You can speed up the process enabling XLA and jit, and disabling CUDA LAUNCH BLOCKING, but may present some memory allocations problems and random freezes. Use the following before creating variables and models.
> 
> ```
> import os
> import tensorflow as tf 
> 
> os.environ['TF_DEVICE_MIN_SYS_MEMORY_IN_MB'] = '1050'
> os.environ['TF_XLA_FLAGS']='--tf_xla_enable_xla_devices --tf_xla_auto_jit=fusible --tf_xla_cpu_global_jit --tf_xla_always_defer_compilation=false --tf_xla_enable_lazy_compilation=true'
> 
> os.environ['TF_ENABLE_GPU_GARBAGE_COLLECTION']='false'
> os.environ['TF_FORCE_GPU_ALLOW_GROWTH']='true'
> 
> os.environ['CUDA_LAUNCH_BLOCKING'] = '0'
> 
> print('jit:',tf.config.optimizer.get_jit())
> tf.config.optimizer.set_jit(True)
> print('jit:',tf.config.optimizer.get_jit())
> ```

I already had XLA enabled for CUDA but I will try with your recommended system variables and report back.

**Comment by taesookim0412** [comment]
![memerror](https://user-images.githubusercontent.com/47997656/105098296-b67cf400-5a5e-11eb-9f2e-0b5f48a1c368.png)
I'm unsure if this is significant but here's what happened at one run.

TF_XLA_FLAGS = --tf_xla_enable_xla_devices --tf_xla_auto_jit=fusible --tf_xla_cpu_global_jit --tf_xla_always_defer_compilation=false --tf_xla_enable_lazy_compilation=true

'TF_DEVICE_MIN_SYS_MEMORY_IN_MB': '1050',
'TF_ENABLE_GPU_GARBAGE_COLLECTION': 'false',
'TF_FORCE_GPU_ALLOW_GROWTH': 'true',
'TF_XLA_FLAGS': '--tf_xla_enable_xla_devices --tf_xla_auto_jit=fusible --tf_xla_cpu_global_jit --tf_xla_always_defer_compilation=false --tf_xla_enable_lazy_compilation=true',

By the way my terminal keeps getting 
tensorflow/core/platform/windows/subprocess.cc:308] SubProcess ended with return code: 0

**Comment by taesookim0412** [comment]
I've tried 3-5+ times trying to get it to run with 
--tf_xla_auto_jit=fusible

however I can't use that flag. It either gives that error or xla fails to compile my functions and it just hangs, eventually closing the kernel.

It yells at me a lot with the subprocess error but I was able to get it to run without any
--tf_xla_auto_jit

flag

`'
'CUDA_LAUNCH_BLOCKING': '0',
TF_DEVICE_MIN_SYS_MEMORY_IN_MB': '1050',
'TF_ENABLE_GPU_GARBAGE_COLLECTION': 'false',
'TF_FORCE_GPU_ALLOW_GROWTH': 'true',
        'TF_XLA_FLAGS': '--tf_xla_enable_xla_devices --tf_xla_cpu_global_jit --tf_xla_always_defer_compilation=false --tf_xla_enable_lazy_compilation=true',`

**Comment by tomerk** [comment]
@Marmotte06 sorry this got lost in my inbox!

The most performant way to turn large numpy array to datasets is to use `from_numpy` from TensorflowIO: https://www.tensorflow.org/io/api_docs/python/tfio/experimental/IODataset#from_numpy

(It avoids the need for memory copies which from_tensor_slices introduces).

From_generator generally isn't very performant because you get bottlenecked by the limitations of python, and don't get the full performance advantages of tf.data.

@taesookim0412 if you haven't done so already, could you open a separate bug and include context around your example, your operating system & drivers, a small code repo if possible, and a profile? As-is what you've reported so far isn't super actionable for us. The comment you quoted itself noted that some of those systems environments can introduce memory allocation issues and crashes, so it's not surprising to me you're running into memory crashing issues when testing them out.

**Comment by Marmotte06** [comment]
Thanks @tomerk , I just had a try with tfio.from_numpy. The good news is that it works fine and is an easy replacement to tf.from_tensor_slices.
However, the time to process an epoch is exactly the same, I see absolutely no difference.
But I need to have another try with one of my huge datasets (60 Gb) which makes tf.from_tensor_slices crash brutally without any error message.
Maybe tfio.from_numpy will luckily pass it...

**Comment by tomerk** [comment]
I imagine tfio.from_numpy will be much better at handling your large datasets. without sharing the actual profiles from your traces though, it's hard to say more about the cause of your performance issues. (If you share traces from the profiler we'd be happy to take a look)

**Comment by tomerk** [comment]
Closing because this is non-actionable for us currently. Feel free to open a new issue (Under a more relevant title) if/when you have profile trace files to share that we can look at in detail.

**Comment by Marmotte06** [comment]
@tomerk , I now confirm that tfio.experimental.IODataset.from_numpy allows to process much larger numpy arrays than tf.data.Dataset.from_tensor_slices or tf.data.Dataset.from_tensors, thanks a lot for the advise.

A couple of questions:
1. In the tfio doc I see there is also a tfio.v0.IODataset, would you know what's the difference with the experimental version ?
2. In the options for the OneDeviceStrategy I am setting experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA . Is that correct, or would AUTO or OFF be better ?

Thanks again
